src/common/DOMWorld.ts

```diff
@@ -464,6 +473,82 @@ export class DOMWorld {
     return queryHandler.waitFor(this, updatedSelector, options);
   }
 
+  /**
+   * @internal
+   */
+  async addBindingToContext(name: string) {
+    const expression = helper.pageBindingInitString('internal', name);
+    try {
+      const context = await this.executionContext();
+      await context._client.send('Runtime.addBinding', {
+        name,
+        executionContextId: context._contextId,
+      });
+      await context.evaluate(expression);
+    } catch (error) {
+      // When the page is navigated, the promise is rejected.
+      // We will try again in the new execution context.
+      if (error.message.includes('Execution context was destroyed')) return;
+      // We could have tried to evaluate in a context which was already
+      // destroyed.
+      if (error.message.includes('Cannot find context with specified id'))
+        return;
+      debugError(error);
+    }
+  }
+
+  /**
+   * @internal
+   */
+  async addBinding(name: string, puppeteerFunction: Function): Promise<void> {
+    if (this._ctxBindings.has(name))
+      throw new Error(
+        `Failed to add page binding with name ${name}: window['${name}'] already exists!`
```

Comment:
Let’s make sure this produces a correct message for arbitrary string values for `name`s (e.g. if `name` includes a `'`):

```suggestion
        `Failed to add page binding with name ${name}: window[${JSON.stringify(name)}] already exists!`
```

Similar to the other comment, we could use `globalThis` instead of `window` — WDYT?

Reply:
Good idea, done.